File: python/graphs/rotting-oranges.py
class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        neighbors = [(0, 1), (1, 0), (0, -1), (-1, 0)]
        duration = 0
        q = deque()
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == 2:
                    q.append((i, j, 0))

        while q:
            current = q.popleft()
            if current[0] < 0 or current[1] < 0 or current[0] >= len(grid) or current[1] >= len(grid[0]) or (
                    grid[current[0]][current[1]] != 2 and grid[current[0]][current[1]] != 1):
                continue

            grid[current[0]][current[1]] = 3
            duration = max(duration, current[2])

            for neighbor in neighbors:
                q.append((current[0] + neighbor[0], current[1] + neighbor[1], current[2] + 1))

        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == 1:
                    return -1

        return duration

File: c++/graphs/rotting-oranges.cc
class Solution {
public:
    int orangesRotting(vector<vector<int>>& grid) {
        std::vector<std::pair<int, int>> neighbors = {std::make_pair(0, 1), std::make_pair(1, 0), std::make_pair(0, -1), std::make_pair(-1, 0)};
        std::queue<std::pair<std::pair<int, int>, int>> q;
        for (int i = 0; i < grid.size(); ++i) {
            for (int j = 0; j < grid[0].size(); ++j) {
                if (grid[i][j] == 2) {
                    q.push(std::make_pair(std::make_pair(i, j), 0));
                }
            }
        }

        int answer = 0;
        while (q.size() > 0) {
            std::pair<std::pair<int, int>, int> element = q.front();
            q.pop();

            if (element.first.first < 0 || element.first.second < 0 || element.first.first >= grid.size() || element.first.second >= grid[0].size() || grid[element.first.first][element.first.second] == 0) {
                continue;
            }

            grid[element.first.first][element.first.second] = 2;
            answer = std::max(answer, element.second);

            for (auto [n_i, n_j] : neighbors) {
                q.push(std::make_pair(std::make_pair(element.first.first + n_i, element.first.second + n_j), element.second + 1));
            }
            grid[element.first.first][element.first.second] = 0;
        }

        for (int i = 0; i < grid.size(); ++i) {
            for (int j = 0; j < grid[0].size(); ++j) {
                if (grid[i][j] == 1) {
                    return -1;
                }
            }
        }
        
        return answer;
    }
};
